File: packages/sdkwork-utils-typescript/src/tests/utils.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import {
  camelCase,
  isBlank,
  slugify,
  truncate,
} from "../string.js";
import { addHours, diffMillis, parseDatetime } from "../datetime.js";
import { base64Encode, hexEncode } from "../encoding.js";
import { deepMerge, getPath, setPath } from "../object.js";
import { hmacSha256, sha256Hash } from "../crypto.js";
import { formatNumberLocale, formatDatetimeLocaleStr } from "../i18n.js";
import { coalesce, defaultIfBlank } from "../optional.js";
import { ok, unwrapOr } from "../result.js";

test("string helpers", () => {
  assert.equal(isBlank("  "), true);
  assert.equal(camelCase("hello_world"), "helloWorld");
  assert.equal(slugify("Hello, SDKWork!"), "hello-sdk-work");
  assert.equal(truncate("abcdef", 5), "ab...");
});

test("datetime helpers", () => {
  const first = parseDatetime("2024-01-01T00:00:00.000Z");
  assert.ok(first);
  const second = addHours(first!, 2);
  assert.equal(diffMillis(first!, second), 7_200_000);
});

test("encoding helpers", () => {
  const bytes = new TextEncoder().encode("hello");
  assert.equal(base64Encode(bytes), "aGVsbG8=");
  assert.equal(hexEncode(bytes), "68656c6c6f");
});

test("object and crypto helpers", () => {
  const merged = deepMerge({ a: 1, nested: { x: 1 } }, { b: 2, nested: { y: 2 } }) as Record<
    string,
    unknown
  >;
  assert.deepEqual(merged, { a: 1, b: 2, nested: { x: 1, y: 2 } });
  const updated = setPath({}, "user.city", "Paris");
  assert.equal(getPath(updated, "user.city"), "Paris");
  assert.equal(
    sha256Hash("hello"),
    "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824",
  );
  assert.equal(
    hmacSha256("payload", "secret"),
    "b82fcb791acec57859b989b430a826488ce2e479fdf92326bd0a2e8375a42ba4",
  );
});

test("optional result and i18n helpers", () => {
  assert.equal(coalesce(null, "", "  ", "ok"), "ok");
  assert.equal(defaultIfBlank("  ", "fallback"), "fallback");
  assert.equal(unwrapOr(ok(42), 0), 42);
  assert.equal(unwrapOr({ ok: false, error: "fail" }, 0), 0);
  assert.equal(formatNumberLocale(1234.5, "en-US", 2), "1,234.50");
  assert.equal(formatNumberLocale(1234.5, "de-DE", 2), "1.234,50");
  assert.ok(formatDatetimeLocaleStr("2024-06-15T14:30:00.000Z", "en-US")?.includes("2024"));
});

File: packages/sdkwork-utils-go/utils_test.go
package utils

import "testing"
import "strings"

func TestStringHelpers(t *testing.T) {
	if !IsBlank(nil) {
		t.Fatal("expected blank nil")
	}
	if CamelCase("hello_world") != "helloWorld" {
		t.Fatal("unexpected camel case")
	}
	if Slugify("Hello, SDKWork!") != "hello-sdk-work" {
		t.Fatal("unexpected slug")
	}
	if Truncate("abcdef", 5, "...") != "ab..." {
		t.Fatal("unexpected truncate")
	}
}

func TestDatetimeHelpers(t *testing.T) {
	first, ok := ParseDatetime("2024-01-01T00:00:00Z", DefaultPattern)
	if !ok {
		t.Fatal("parse failed")
	}
	second := AddHours(first, 2)
	if DiffMillis(first, second) != 7_200_000 {
		t.Fatal("unexpected diff")
	}
}

func TestEncodingHelpers(t *testing.T) {
	data := []byte("hello")
	if Base64Encode(data) != "aGVsbG8=" {
		t.Fatal("unexpected base64")
	}
	if HexEncode(data) != "68656c6c6f" {
		t.Fatal("unexpected hex")
	}
}

func TestOptionalResultAndI18nHelpers(t *testing.T) {
	okValue := Coalesce(nil, ptr(""), ptr("  "), ptr("ok"))
	if okValue == nil || *okValue != "ok" {
		t.Fatal("unexpected coalesce")
	}
	if DefaultIfBlank(ptr("  "), "fallback") != "fallback" {
		t.Fatal("unexpected default_if_blank")
	}
	if UnwrapOr(Ok(42), 0) != 42 || UnwrapOr(Err[int]("fail"), 0) != 0 {
		t.Fatal("unexpected unwrap_or")
	}
	if FormatNumberLocale(1234.5, "en-US", 2) != "1,234.50" {
		t.Fatal("unexpected en-US number locale")
	}
	if FormatNumberLocale(1234.5, "de-DE", 2) != "1.234,50" {
		t.Fatal("unexpected de-DE number locale")
	}
	formatted, ok := FormatDatetimeLocaleStr("2024-06-15T14:30:00.000Z", "en-US")
	if !ok || !strings.Contains(formatted, "2024") {
		t.Fatal("unexpected datetime locale")
	}
}

func ptr(value string) *string {
	return &value
}

func TestObjectAndCryptoHelpers(t *testing.T) {
	merged := DeepMerge(
		map[string]any{"a": 1, "nested": map[string]any{"x": 1}},
		map[string]any{"b": 2, "nested": map[string]any{"y": 2}},
	).(map[string]any)
	nested := merged["nested"].(map[string]any)
	if nested["x"] != 1 || nested["y"] != 2 {
		t.Fatal("unexpected deep merge")
	}
	updated := SetPath(map[string]any{}, "user.city", "Paris")
	city, ok := GetPath(updated, "user.city")
	if !ok || city != "Paris" {
		t.Fatal("unexpected set/get path")
	}
	if Sha256Hash([]byte("hello")) != "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824" {
		t.Fatal("unexpected sha256")
	}
	if HmacSha256([]byte("payload"), []byte("secret")) != "b82fcb791acec57859b989b430a826488ce2e479fdf92326bd0a2e8375a42ba4" {
		t.Fatal("unexpected hmac")
	}
}
